pycc/ccwfn.py

```diff
@@ -0,0 +1,352 @@
+    H : Hamiltonian object
+        the normal-ordered Hamiltonian, which includes the Fock matrix, the ERIs, and the spin-adapted ERIs (L)
+    o : NumPy slice
+        occupied orbital subspace
+    v : NumPy slice
+        virtual orbital subspace
+    Dia : NumPy array
+        one-electron energy denominator
+    Dijab : NumPy array
+        two-electron energy denominator
+    t1 : NumPy array
+        T1 amplitudes
+    t2 : NumPy array
+        T2 amplitudes
+    ecc | float
+        the final CC correlation energy
+
+    Methods
+    -------
+    solve_ccsd()
+        Solves the CCSD T amplitude equations
+    residuals()
+        Computes the T1 and T2 residuals for a given set of amplitudes and Fock operator
+    """
+
+    def __init__(self, scf_wfn, **kwargs):
+        """
+        Parameters
+        ----------
+        scf_wfn : Psi4 Wavefunction Object
+            computed by Psi4 energy() method
+
+        Returns
+        -------
+        None
+        """
+
+        time_init = time.time()
+
+        valid_cc_models = ['CCD', 'CC2', 'CCSD', 'CCSD(T)', 'CC3']
+        model = kwargs.pop('model','CCSD')
+        if model not in valid_cc_models:
+            raise Exception("%s is not an allowed CC model." % (model))
+        self.model = model
+
+        # models requiring singles
+        self.need_singles = ['CCSD', 'CCSD(T)']
+
+        # models requiring T1-transformed integrals
+        self.need_t1_transform = ['CC2', 'CC3']
+
+        valid_local_models = [None, 'LPNO']
+        local = kwargs.pop('local', None)
+        if local not in valid_local_models:
+            raise Exception("%s is not an allowed local-CC model." % (local))
+        self.local = local
+        self.local_cutoff = kwargs.pop('lpno_cutoff', 1e-5)
+
+        valid_local_MOs = ['PIPEK_MEZEY', 'BOYS']
+        local_MOs = kwargs.pop('local_mos', 'PIPEK_MEZEY')
+        if local_MOs not in valid_local_MOs:
+            raise Exception("%s is not an allowed MO localization method." % (local_MOs))
+        self.local_MOs = local_MOs
+
+        self.ref = scf_wfn
+        self.eref = self.ref.energy()
+        self.nfzc = self.ref.frzcpi()[0]                # assumes symmetry c1
+        self.no = self.ref.doccpi()[0] - self.nfzc      # active occ; assumes closed-shell
+        self.nmo = self.ref.nmo()                       # all MOs/AOs
+        self.nv = self.nmo - self.no - self.nfzc        # active virt
+        self.nact = self.no + self.nv                   # all active MOs
+
+        print("NMO = %d; NACT = %d; NO = %d; NV = %d" % (self.nmo, self.nact, self.no, self.nv))
+
+        # orbital subspaces
+        self.o = slice(0, self.no)
+        self.v = slice(self.no, self.nmo)
+
+        # For convenience
+        o = self.o
+        v = self.v
+
+        # Get MOs
+        C = self.ref.Ca_subset("AO", "ACTIVE")
+        npC = np.asarray(C)  # as numpy array
+        self.C = C
+
+        # Localize occupied MOs if requested
```

Comment:
the occupied-space localization feels a little out of place inside `ccwfn`, but having it here keeps us from having to remember to reset `C` in all of the other objects to match the `Hamiltonian` object. seems fair.

Reply:
`ccwfn` is the top-level class from which the others are derived.  I admit that this is strange because, in some sense, the `hamiltonian` ought to be at top level, or at least a friend class to `ccwfn`, but I didn't want the user to have to instantiate the `hamiltonian` separately.  Thus, the MOs fit most naturally into `ccwfn` in this structure, in my opinion.